packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp

```diff
@@ -205,109 +247,77 @@ preEvaluate(typename TRAITS::PreEvalData d)
-      Teuchos::ArrayRCP<double> local_dc;
-      for(std::size_t fieldIndex = 0; fieldIndex < scatterFields_.size(); fieldIndex++) {
-         int fieldNum = fieldIds_[fieldIndex];
-         int indexerId = globalIndexer_->getFieldBlock(fieldNum);
-
-         RCP<SpmdVectorBase<double> > dc = rcp_dynamic_cast<SpmdVectorBase<double> >(dirichletCounter_->getNonconstVectorBlock(indexerId));
-         dc->getNonconstLocalData(ptrFromRef(local_dc));
-
-         // grab local data for inputing
-         RCP<SpmdVectorBase<double> > block_r = rcp_dynamic_cast<SpmdVectorBase<double> >(r->getNonconstVectorBlock(indexerId));
-         block_r->getNonconstLocalData(ptrFromRef(local_r));
-
-         if (!scatterIC_) {
-           // this call "should" get the right ordering according to the Intrepid2 basis
-           const std::pair<std::vector<int>,std::vector<int> > & indicePair 
-             = globalIndexer_->getGIDFieldOffsets_closure(blockId,fieldNum, side_subcell_dim_, local_side_id_);
-           const std::vector<int> & elmtOffset = indicePair.first;
-           const std::vector<int> & basisIdMap = indicePair.second;
-
-           // loop over basis functions
-           for(std::size_t basis=0;basis<elmtOffset.size();basis++) {
-             int offset = elmtOffset[basis];
-             int lid = LIDs[offset];
-             if(lid<0) // not on this processor!
+   // Loop over scattered fields
+   int currentWorksetLIDSubBlock = -1;
+   for (std::size_t fieldIndex = 0; fieldIndex < scatterFields_.size(); fieldIndex++) {
+     // workset LIDs only change for different sub blocks
+     if (productVectorBlockIndex_[fieldIndex] != currentWorksetLIDSubBlock) {
+       fieldGlobalIndexers_[fieldIndex]->getElementLIDs(localCellIds,worksetLIDs_);
+       currentWorksetLIDSubBlock = productVectorBlockIndex_[fieldIndex];
+     }
+
+     // Get Scatter target block
+     const auto& tpetraScatterTarget = *((rcp_dynamic_cast<Thyra::TpetraVector<RealType,LO,GO,NodeT>>(thyraScatterTarget->getNonconstVectorBlock(productVectorBlockIndex_[fieldIndex]),true))->getTpetraVector());
+     const auto& kokkosScatterTarget = tpetraScatterTarget.template getLocalView<PHX::mem_space>();
+
+     // Get dirichlet counter block
+     const auto& tpetraDirichletCounter = *((rcp_dynamic_cast<Thyra::TpetraVector<RealType,LO,GO,NodeT>>(dirichletCounter_->getNonconstVectorBlock(productVectorBlockIndex_[fieldIndex]),true))->getTpetraVector());
+     const auto& kokkosDirichletCounter = tpetraDirichletCounter.template getLocalView<PHX::mem_space>();
+
+     // Class data fields for lambda capture
+     const auto& fieldOffsets = fieldOffsets_[fieldIndex];
+     const auto& basisIndices = basisIndexForMDFieldOffsets_[fieldIndex];
+     const auto& worksetLIDs = worksetLIDs_;
+     const auto& fieldValues = scatterFields_[fieldIndex];
+     const auto& applyBC = applyBC_[fieldIndex].get_static_view();
+     const bool checkApplyBC = checkApplyBC_;
```

Comment:
What's the motivation for this block?  Can we not access member data within a `Kokkos::parallel_for`?

Reply:
not on cuda if using lambda capture

Reply:
Oh - should also mention - when we can jump Trilinos to c++17 this should be fixed.

Reply:
Lambda capture works differently in C++17?  I'll set myself a reminder for a week before I retire&hellip;

Reply:
The kokkos team successfully got support for *this capture into the 17 standard:

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0018r3.html